### PR TITLE
fix(lambda): Function allows specifying vpcSubnets without vpc

### PR DESCRIPTION
**Pull request by laurelmay**
This is almost certainly never something that a user would intend.
As-is, `vpcSubnets` does nothing if `vpc` is unspecified, so the Lambda
Function would have no interfaces placed in a VPC. This is unlikely to be what
a user would have expected if they passed a set of private subnets. This
should be an error to prevent users from accidentally thinking they've
put a Lambda Function into a VPC.

If throwing an Error is a breaking change and not acceptable, then
perhaps at least a Warning Annotation should be added in this case.

This is motivated by the feedback received during review of #21357.

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 1]
```diff
@@ -680,6 +680,8 @@ export class Function extends FunctionBase {
     if (props.vpc) {
       // Policy that will have ENI creation permissions
       managedPolicies.push(iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaVPCAccessExecutionRole'));
+    } else if (props.vpcSubnets) {
```
Comment (corymhall) [inline packages/@aws-cdk/aws-lambda/lib/function.ts:683]: I think I'm fine with this change given that it does not cause any changes to the deployed infra.

One nit - can you put this check in the `configureVpc` method? It feels like it should go there since that is where the other checks are.

**Review by corymhall — CHANGES_REQUESTED** [review]

**Comment by kylelaker** [comment]
Hoping the failure on ec5cd4f is fixed by #21374 and therefore the merge from `main`.

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 2]
```diff
@@ -1094,7 +1094,12 @@ Environment variables can be marked for removal when used in Lambda@Edge by sett
       throw new Error('Cannot configure \'securityGroup\' or \'allowAllOutbound\' without configuring a VPC');
     }
 
-    if (!props.vpc) { return undefined; }
+    if (!props.vpc) {
+      if (props.vpcSubnets) {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-lambda/lib/function.ts:1098]:
```suggestion
    if (!props.vpc && props.vpcSubnets) {
```

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 3]
```diff
@@ -1094,7 +1094,12 @@ Environment variables can be marked for removal when used in Lambda@Edge by sett
...
+        throw new Error("'vpc' must be specified with 'vpcSubnets' in order to place the Function's interfaces in a VPC");
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-lambda/lib/function.ts:1099]:
```suggestion
        throw new Error("'vpcSubnets' may not be specified without 'vpc'");
```

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Just a couple of stylistic suggestions.

File: packages/@aws-cdk/aws-lambda/lib/function.ts (hunk 2 above)
Comment (kylelaker) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:1098]: I thought about doing this, but there's the `return undefined;` after the inner `if`. So it'd have to be rewritten as:

```ts
if (!props.vpc && props.vpcSubnets) {
  throw new Error("...");
} else if (!props.vpc) {
  return undefined
}
```

which felt a little messy and maybe more confusing to repeat half of the condition?

Actually mean to include a note/question about this when I pushed the new commit but got distracted by the failing build 😄 So I'm happy to making the change but just want to make sure that this is the desired outcome?

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:1098]: No need for you to revise because of a small stylistic difference. I only made the suggestion because I was also making the suggestion to change the error message. Otherwise, I would have just plain approved since @corymhall is also in favor.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:1098]: But also, I didn't notice that the error message above was infinitely better than the one I suggested. So, sorry, one more suggested change to that.

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 4]
```diff
@@ -1094,7 +1094,12 @@ Environment variables can be marked for removal when used in Lambda@Edge by sett
       throw new Error('Cannot configure \'securityGroup\' or \'allowAllOutbound\' without configuring a VPC');
     }
 
-    if (!props.vpc) { return undefined; }
+    if (!props.vpc) {
+      if (props.vpcSubnets) {
+        throw new Error("'vpcSubnets' may not be specified without 'vpc'");
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-lambda/lib/function.ts:1099]:
```suggestion
        throw new Error('Cannot configure \'vpcSubnets\' without configuring a VPC');
```

File: packages/@aws-cdk/aws-lambda/test/vpc-lambda.test.ts [hunk 5]
```diff
@@ -323,6 +323,21 @@ describe('lambda + vpc', () => {
...
+
+    // WHEN
+    expect(() => {
+      new lambda.Function(stack, 'Function', {
+        code: new lambda.InlineCode('foo'),
+        handler: 'index.handler',
+        runtime: lambda.Runtime.NODEJS_14_X,
+        vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE },
+      });
+    }).toThrow("'vpcSubnets' may not be specified without 'vpc'");
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-lambda/test/vpc-lambda.test.ts:339]:
```suggestion
    }).toThrow('Cannot configure \'vpcSubnets\' without configuring a VPC');
```

**Review by TheRealAmazonKendra — APPROVED** [review]
I should have looked at the code directly above and made my suggestion to align to that. Sorry for the extra back and forth!

**Review by TheRealAmazonKendra — DISMISSED** [review]
Aaand then I hit the wrong button...

File: packages/@aws-cdk/aws-lambda/lib/function.ts (hunk 2 above)
Comment (kylelaker) [reply packages/@aws-cdk/aws-lambda/lib/function.ts:1098]: All good! And I meant "meant to" instead of "mean to" -- oops, sorta changes the tone.  But in any case, thanks for the feedback on the error message. I read https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#error-messages and I think ended up writing something too wordy; this one is much clearer!

**Comment by kylelaker** [comment]
Not sure what's up with the CodeBuild failure this time around -- it doesn't really feel related to anything that's changed and running the `integ-runner` tests locally doesn't result in a failure. I think I'm just going to hope it's a one-off issue and that re-triggering a build fixes it? But I don't know how to ask CodeBuild to rebuild without a new commit so I'll just click the Update branch button? 🤷🏻  

```
@aws-cdk/integ-runner: FAIL test/runner/snapshot-test-runner.test.js
@aws-cdk/integ-runner:   ● IntegTest runSnapshotTests › with defaults no diff
@aws-cdk/integ-runner:     Cannot read integ manifest 'test/test-data/test-with-snapshot.integ.snapshot/manifest.json': Unexpected end of JSON input
@aws-cdk/integ-runner:       32 |
@aws-cdk/integ-runner:       33 |     } catch (e) {
@aws-cdk/integ-runner:     > 34 |       throw new Error(`Cannot read integ manifest '${fileName}': ${e.message}`);
@aws-cdk/integ-runner:          |             ^
@aws-cdk/integ-runner:       35 |     }
@aws-cdk/integ-runner:       36 |   }
@aws-cdk/integ-runner:       37 |
@aws-cdk/integ-runner:       at Function.fromFile (lib/runner/private/cloud-assembly.ts:34:13)
@aws-cdk/integ-runner:       at Function.fromPath (lib/runner/private/cloud-assembly.ts:51:37)
@aws-cdk/integ-runner:       at IntegSnapshotRunner.readAssembly (lib/runner/snapshot-test-runner.ts:223:45)
@aws-cdk/integ-runner:       at IntegSnapshotRunner.testSnapshot (lib/runner/snapshot-test-runner.ts:55:33)
@aws-cdk/integ-runner:       at Object.<anonymous> (test/runner/snapshot-test-runner.test.ts:55:31)
```

**Review by TheRealAmazonKendra — APPROVED** [review]
